src/transform/src/monotonic.rs

```diff
@@ -120,38 +123,84 @@ impl MonotonicFlag {
-                        }
+                        let inserted = locals.insert(id.clone());
+                        assert!(inserted, "Shadowing of identifier: {id:?}");
                     }
-                    // Pessimistically assume all bindings are non-monotonic, and
-                    // iteratively attempt to determine that some may be monotonic.
+
+                    // Sequentially AND locals[id] with the result of descending
+                    // into a clone of values[id]. Repeat until one of the
+                    // following conditions is met:
                     //
-                    // TODO: This is suboptimal as long as there is one method to
-                    // both report and apply monotonicity information. We need to
-                    // be able to repeatedly report monotonicity information, and
-                    // then apply it subsequently. Ideally we would optimistically
-                    // assume all bindings were monotonic, and iteratively remove
-                    // any that are determined to be non-monotonic; only once this
-                    // concludes can we apply any transformations, however.
-                    // Don't forget to handle `max_iters` when changing to the
-                    // optimistic approach!
-                    let mut added = true;
-                    while added {
-                        added = false;
-                        for (id, value) in ids.iter().zip(&mut *values) {
-                            if !locals.contains(id) && self.apply(value, mon_ids, locals)? {
-                                added = true;
-                                locals.insert(*id);
+                    // 1. The locals entries have stabilized at a fixpoint.
+                    // 2. No fixpoint was found after `max_iterations`. If this
+                    //    is the case reset the locals entries for all recursive
+                    //    CTEs to `false` (pessimistic approximation).
+                    // 3. We reach the user-specified iteration limit of any of
+                    //    the bindings. In this case, we also give up similarly
+                    //    to (2), because we don't want to complicate things with
+                    //    handling different limits per binding.
+                    let min_max_iter = max_iters.into_iter().filter_map(|md| *md).min();
+                    let max_iterations = 100;
+                    let mut curr_iteration = 0;
+                    loop {
+                        // Check for conditions (2) and (3).
+                        if curr_iteration > max_iterations
+                            || min_max_iter
+                                .map(|min_max_iter| curr_iteration > min_max_iter.get())
+                                .unwrap_or(false)
+                        {
+                            if curr_iteration > u64::cast_from(ids.len()) {
```

Comment:
The conditions here use `>` because I was hitting some of the panics below with trivial queries (where `ids.len()` was 1), whereas the ones in `ColumnKnowledge` use `>=`. Maybe I should `ColumnKnowledge` as well in this PR.

Reply:
So, there are 3 conditions here:
- The panic one (`curr_iteration > u64::cast_from(ids.len())`) definitely has to be `>`, because we need one extra iteration after the analysis reached the final value to detect the convergence.
- `curr_iteration > max_iterations` doesn't matter whether it's `>` or `>=`, because `max_iterations` being 100 doesn't have any actual reason for being exactly 100, it is just a heuristically determined value.
- For `curr_iteration > min_max_iter.get()`, it's slightly better to have `>=`, but `>` won't cause a correctness issue. In more detail, if we have `>`, then we might do one more iteration than what was specified in the limit setting. This can make the final result stricter (less monotonic) than the ideal value, but that's fine from a correctness perspective, because we are looking for a lower bound on the ideal value. In theory, it might prevent some valid optimization from happening, though.

`ColumnKnowledge` looks fine to me, in light of the above. (I think it would be good to set the `curr_iteration > max_iterations` here consistently with `ColumnKnowledge`, i.e., to `>=`.)